master/custom_steps.py

```diff
@@ -1,10 +1,130 @@
+import os
+import re
 import xml.etree.ElementTree as Xml
+from collections import defaultdict
+from pathlib import Path
+from typing import Dict, List
 
 from buildbot.process.buildstep import BuildStepFailed, BuildStep, ShellMixin
+from buildbot.process.results import SUCCESS, FAILURE, WARNINGS
 from buildbot.steps.worker import CompositeStepMixin
 from twisted.internet import defer
 
-__all__ = ['CTest']
+__all__ = ['CleanOldFiles', 'CTest', 'SetPropertiesFromCMakeCache']
+
+
+class SetPropertiesFromCMakeCache(CompositeStepMixin, BuildStep):
+    name = 'set-properties-from-cmake-cache'
+
+    renderables = ['props']
+
+    _cache_re = re.compile(
+        r'''
+        ^(?!//|\#)    # Ignore comment lines.
```

Comment:
...even with all these detailed comments, this fills me with terror at future maintenance issues. Is the format of CMakeCache.txt well-defined and suitable for mining in this way? Is this really the best way to transmit the version number? Is there no more-explicit way we could do it that wouldn't rely on plumbing into CMake's implementation details?

Reply:
> Is the format of CMakeCache.txt well-defined and suitable for mining in this way?

The format of CMakeCache.txt has remained stable for over 15 years. Looking at git blame shows that the code for parsing it was moved from one file to another 5 years ago and was last actually modified between 15-20 years ago. So I'm not worried about them changing this without at least signalling a major version bump to CMake 4.x

> Is this really the best way to transmit the version number? Is there no more-explicit way we could do it that wouldn't rely on plumbing into CMake's implementation details?

Not without patching and backporting a change to our CMakeLists.txt. The issue is that while newer versions of CMake (3.12+) write `CMAKE_PROJECT_VERSION` to the cache, they do so as a `STATIC`-type variable. Because the `cmake -L` cache-listing tool is ridiculous, it excludes `STATIC` and `INTERNAL` entries.

Ben Boeckel (a maintainer) just told me to grep on the Discourse: https://discourse.cmake.org/t/read-project-version-from-cache/1849

The backporting option is to save a `STRING`-type variable called `Halide_VERSION` to the cache and use the command line tool to read it (which will involve parsing with the same regex, because that's the format it provides).

Reply:
Bleah. OK Then. May I suggest adding (some) of this explanation to script comments for future reviewers.

Reply:
Also, I coded the step so that if the expected variable(s) are not found, the step returns a warning status and logs, so we can at least respond the _first_ time something breaks.